### PR TITLE
Add documentation to several classes

### LINKED ISSUE #2764: API docs for publicly exposed modules/classes
I'm currently looking into how to improve the addon docs and while my initial idea was to put it in a guide-like setup, I'm now starting to think this should probably be more of an API reference.

Some of the addon hooks receive objects that are constructed by ember-cli internals (like `options` for `serverMiddleware`, `result` for `postBuild`). It would be great if we could abstract these into classes (eg. a `ServerOptions` class) and then build API docs on all classes that are publicly exposed. Addon docs would then be as simple as generating docs for the `Addon` class.

This would be a better approach then I proposed in #2300 IMO. What do you think @stefanpenner @rwjblue  

**Keeping track of this in #2835**

### PR DESCRIPTION
**Pull request by joostdevries**
Opening a PR for this so we can discuss the docblocks. Fixes #2764.
Build result updated frequently at http://joostdevries.github.io/ember-cli/api-docs/build/

## Todo:
- [x] Classes
  - [x] Addon
  - [x] EmberApp
  - [x] EmberAddon
  - [x] Project
- [x] Document `CoreObject` (in stefanpenner/core_object)
- [x] Provide links to external docs for..
  - [x] Broccoli trees

### CODE REVIEW

**Comment by joostdevries** [comment]
@stefanpenner This should be good for another round of feedback. Preview still at http://joostdevries.github.io/ember-cli/api-docs/build/

File: lib/broccoli/ember-app.js [hunk 1]
```diff
@@ -212,6 +229,11 @@ function EmberApp(options) {
   this._notifyAddonIncluded();
 }
 
+/**
+  Returns the environment name
+
+  @return {String} Environment name
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: Should indicate that this is a class method (and does not require instantiation), via `@static` YUIDoc flag.

File: lib/broccoli/ember-app.js [hunk 2]
```diff
@@ -251,6 +273,9 @@ EmberApp.prototype._notifyAddonIncluded = function() {
 };
 
 /**
+  Loads and initializes addons for this project.
+  Calls initializeAddons on the Project.
+
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: @private

File: lib/broccoli/ember-app.js (hunk 1 above)
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: Also `@public`

File: lib/broccoli/ember-app.js [hunk 3]
```diff
@@ -258,10 +283,11 @@ EmberApp.prototype.initializeAddons = function() {
...
 
 /**
-  @method addonTreesFor
-  @param type
-  @return
-*/
+  Returns a list of trees for a given type, returned by all addons.
+
+  @param  {String} type Type of tree
+  @return {Array}       List of trees
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: `@private`

File: lib/broccoli/ember-app.js [hunk 4]
```diff
@@ -271,11 +297,13 @@ EmberApp.prototype.addonTreesFor = function(type) {
...
+  Runs addon postprocessing on a given tree and returns the processed tree.
+
   @method addonPostprocessTree
-  @param type
-  @param tree
-  @return
-*/
+  @param  {String} type Type of tree
+  @param  {Tree}   tree Tree to process
+  @return {Tree}        Processed tree
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: `@private`

File: lib/broccoli/ember-app.js [hunk 5]
```diff
@@ -289,6 +317,8 @@ EmberApp.prototype.addonPostprocessTree = function(type, tree) {
 };
 
 /**
+  Imports legacy imports in this.vendorFiles
+
   @method populateLegacyFiles
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: `@private`

File: lib/broccoli/ember-app.js [hunk 6]
```diff
@@ -303,8 +333,10 @@ EmberApp.prototype.populateLegacyFiles = function () {
 };
 
 /**
+  Returns the tree for app/index.html
+
   @method index
-  @return
+  @return {Tree} Tree for app/index.html
```
Comment (rwjblue) [inline lib/broccoli/ember-app.js]: `@private`

**Comment by stefanpenner** [comment]
@joostdevries can we get a rebase good sir?

**Comment by joostdevries** [comment]
@stefanpenner You have gotten a squashed merge instead kind sir.

**Comment by stefanpenner** [comment]
Haha. I'll take it!

**Comment by joostdevries** [comment]
I'll keep updating http://joostdevries.github.io/ember-cli/api-docs/build/ with master by the way

**Pull request merged**